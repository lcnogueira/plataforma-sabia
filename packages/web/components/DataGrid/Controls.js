import React, { useState, useRef, useMemo } from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';
import { useTranslation } from 'react-i18next';
import { useTheme } from 'styled-components';
import {
	GridControls,
	ArrowLeftIcon,
	ArrowRightIcon,
	DoubleArrowLeftIcon,
	DoubleArrowRightIcon,
	AscOrderIcon,
	DescOrderIcon,
} from './styles';
import bookmarksEnum from '../../utils/enums/bookmarks.enum';

const Controls = ({
	data,
	sortOptions,
	handleSortBy,
	currentOrder,
	handlePagination,
	currentPage,
	totalPages,
	totalItems,
	itemsPerPage,
}) => {
	const [sortOption, setSortOption] = useState('');
	const isSelectFocusedRef = useRef(false);
	const { t } = useTranslation(['datagrid']);
	const {
		colors: { lightGray },
	} = useTheme();

	/**
	 * Formats page items count based on current page
	 *
	 * @returns {string} Formatted pagination data, e.g. `1-10 of 20`
	 */
	const getFirstAndLastItem = () => {
		if (currentPage === 1) {
			return {
				firstItem: 1,
				lastItem: data.length,
			};
		}

		if (currentPage === totalPages) {
			return {
				firstItem: totalItems - data.length + 1,
				lastItem: totalItems,
			};
		}

		return {
			firstItem: (currentPage - 1) * itemsPerPage + 1,
			lastItem: data.length + (currentPage - 1) * itemsPerPage,
		};
	};

	const getItemsCountByPage = () => {
		const { firstItem, lastItem } = getFirstAndLastItem();

		return `${firstItem} - ${lastItem} ${t('paginationLabel')} ${totalItems}`;
	};

	const sortByI18 = t('sortByOption');
	const customSelectStyles = useMemo(
		() => ({
			container: (base, state) => {
				isSelectFocusedRef.current = state.isFocused;
				return {
					...base,
					display: 'inline-block',
				};
			},
			placeholder: (base, state) => {
				return {
					...base,
					...(isSelectFocusedRef.current && state.value
						? {}
						: {
								position: 'static',
								top: 'auto',
								transform: 'none',
						  }),
				};
			},
			input: (base, state) => {
				return {
					...base,
					...(isSelectFocusedRef.current && state.value
						? {}
						: {
								position: 'absolute',
						  }),
				};
			},
			singleValue: (base, state) => {
				return {
					...base,
					maxWidth: 'none',
					...(isSelectFocusedRef.current && state.value
						? {}
						: {
								position: 'static',
								top: 'auto',
								transform: 'none',
						  }),
				};
			},
			valueContainer: (base) => ({
				...base,
				alignItems: 'center',
				display: 'flex',

				':before': {
					content: `"${sortByI18}"`,
					color: lightGray,
				},
			}),
		}),
		[sortByI18, lightGray],
	);

	return (
		<GridControls>
			{handleSortBy && (
				<div className="sort-by">
					<Select
						instanceId="bookmarks-ordering-select"
						options={sortOptions}
						styles={customSelectStyles}
						value={sortOption}
						placeholder={t('sortByPlaceholder')}
						onChange={(option) => {
							setSortOption(option);
							handleSortBy(option.value);
						}}
						isSearchable={false}
					/>

					{sortOption && (
						<div className="sort-order">
							<button
								type="button"
								aria-label="Ascending Order"
								onClick={() =>
									handleSortBy(sortOption.value, bookmarksEnum.ASC_ORDER)
								}
							>
								<AscOrderIcon
									className={
										currentOrder === bookmarksEnum.ASC_ORDER
											? 'active'
											: undefined
									}
								/>
							</button>
							<button
								type="button"
								aria-label="Descending Order"
								onClick={() =>
									handleSortBy(sortOption.value, bookmarksEnum.DESC_ORDER)
								}
							>
								<DescOrderIcon
									className={
										currentOrder === bookmarksEnum.DESC_ORDER
											? 'active'
											: undefined
									}
								/>
							</button>
						</div>
					)}
				</div>
			)}

			{handlePagination && (
				<div className="pagination">
					<button
						type="button"
						onClick={() => handlePagination(1)}
						disabled={currentPage === 1}
						aria-label="First Page"
					>
						<DoubleArrowLeftIcon />
					</button>
					<button
						type="button"
						onClick={() => handlePagination(currentPage - 1)}
						disabled={currentPage === 1}
						aria-label="Previous Page"
					>
						<ArrowLeftIcon />
					</button>
					<span>{getItemsCountByPage()}</span>
					<button
						type="button"
						onClick={() => handlePagination(currentPage + 1)}
						disabled={currentPage === totalPages}
						aria-label="Next Page"
					>
						<ArrowRightIcon />
					</button>
					<button
						type="button"
						onClick={() => handlePagination(totalPages)}
						disabled={currentPage === totalPages}
						aria-label="Last Page"
					>
						<DoubleArrowRightIcon />
					</button>
				</div>
			)}
		</GridControls>
	);
};

Controls.propTypes = {
	data: PropTypes.arrayOf(
		PropTypes.shape({
			id: PropTypes.number,
		}),
	).isRequired,
	sortOptions: PropTypes.arrayOf(
		PropTypes.shape({
			value: PropTypes.string,
			label: PropTypes.string,
		}),
	),
	handleSortBy: PropTypes.func,
	currentOrder: PropTypes.string,
	handlePagination: PropTypes.func,
	currentPage: PropTypes.number,
	totalPages: PropTypes.number,
	totalItems: PropTypes.number,
	itemsPerPage: PropTypes.number,
};

Controls.defaultProps = {
	sortOptions: [],
	handleSortBy: null,
	currentOrder: '',
	handlePagination: null,
	currentPage: 1,
	totalPages: 1,
	totalItems: 1,
	itemsPerPage: 1,
};

export default Controls;
