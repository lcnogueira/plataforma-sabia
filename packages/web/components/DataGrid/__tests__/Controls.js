import React from 'react';
import { render, fireEvent } from 'test-utils';

import Controls from '../Controls';

// Component props mock
const mockData = [];
const mockSortOptions = [
	{ value: 'id', label: 'Id' },
	{ value: 'title', label: 'TÃ­tulo' },
	{ value: 'status', label: 'Status' },
];
let mockHandleSortBy;
let mockHandlePagination;

jest.mock('react-select', () => ({ options, value, onChange }) => {
	function handleChange(event) {
		const option = options.find((opt) => opt.value === event.currentTarget.value);

		onChange(option);
	}

	return (
		<select data-testid="select" value={value} onChange={handleChange}>
			{options.map(({ label, value: oValue }) => (
				<option key={label} value={oValue}>
					{label}
				</option>
			))}
		</select>
	);
});

describe('<Controls />', () => {
	beforeEach(() => {
		mockHandleSortBy = jest.fn((option, order) => ({ option, order }));
		mockHandlePagination = jest.fn((newPage) => newPage);
	});

	it('should render without crashing', () => {
		const { container } = render(<Controls data={mockData} />);
		expect(container).toMatchSnapshot();
	});

	it('should render order select if handleSortBy is not null', () => {
		const { getByTestId } = render(
			<Controls
				data={mockData}
				handleSortBy={mockHandleSortBy}
				sortOptions={mockSortOptions}
			/>,
		);

		expect(getByTestId('select')).toBeInTheDocument();
	});

	it('should render ascending and descending buttons if select value is not empty', () => {
		const { getByLabelText, getByTestId } = render(
			<Controls
				data={mockData}
				handleSortBy={mockHandleSortBy}
				sortOptions={mockSortOptions}
			/>,
		);

		fireEvent.change(getByTestId('select'), {
			target: { value: 'title' },
		});

		expect(getByLabelText('Ascending Order')).toBeInTheDocument();
		expect(getByLabelText('Descending Order')).toBeInTheDocument();
	});

	it('should exec handleSortBy passing option as first argument and undefined as second argument when changing select ', () => {
		const { getByTestId } = render(
			<Controls
				data={mockData}
				handleSortBy={mockHandleSortBy}
				sortOptions={mockSortOptions}
			/>,
		);

		fireEvent.change(getByTestId('select'), {
			target: { value: 'title' },
		});

		expect(mockHandleSortBy.mock.calls).toHaveLength(1);
		// The first argument of the first call to the function should be 'title'
		expect(mockHandleSortBy.mock.calls[0][0]).toBe('title');
		// The second argument of the first call to the function should be 'undefined'
		expect(mockHandleSortBy.mock.calls[0][1]).toBeUndefined();
	});

	it('should exec handleSortBy passing current option as first argument and `ASC` as second argument when clicking ascending button ', () => {
		const { getByLabelText, getByTestId } = render(
			<Controls
				data={mockData}
				handleSortBy={mockHandleSortBy}
				sortOptions={mockSortOptions}
			/>,
		);

		fireEvent.change(getByTestId('select'), {
			target: { value: 'status' },
		});

		fireEvent.click(getByLabelText('Ascending Order'));

		expect(mockHandleSortBy.mock.calls).toHaveLength(2);
		// The first argument of the second call to the function should be 'status'
		expect(mockHandleSortBy.mock.calls[1][0]).toBe('status');
		// The second argument of the second call to the function should be 'ASC'
		expect(mockHandleSortBy.mock.calls[1][1]).toBe('ASC');
	});

	it('should exec handleSortBy passing current option as first argument and `DESC` as second argument when clicking descending button ', () => {
		const { getByLabelText, getByTestId } = render(
			<Controls
				data={mockData}
				handleSortBy={mockHandleSortBy}
				sortOptions={mockSortOptions}
			/>,
		);

		fireEvent.change(getByTestId('select'), {
			target: { value: 'id' },
		});

		fireEvent.click(getByLabelText('Descending Order'));

		expect(mockHandleSortBy.mock.calls).toHaveLength(2);
		// The first argument of the second call to the function should be 'id'
		expect(mockHandleSortBy.mock.calls[1][0]).toBe('id');
		// The second argument of the second call to the function should be 'DESC'
		expect(mockHandleSortBy.mock.calls[1][1]).toBe('DESC');
	});

	it('should render pagination controls if handlePagination is not null', () => {
		const { getByLabelText } = render(
			<Controls data={mockData} handlePagination={mockHandlePagination} />,
		);

		expect(getByLabelText('First Page')).toBeInTheDocument();
		expect(getByLabelText('Previous Page')).toBeInTheDocument();
		expect(getByLabelText('Next Page')).toBeInTheDocument();
		expect(getByLabelText('Last Page')).toBeInTheDocument();
	});

	it('should disable pagination controls if totalPages equals 1', () => {
		const { getByLabelText } = render(
			<Controls data={mockData} handlePagination={mockHandlePagination} />,
		);

		expect(getByLabelText('First Page')).toBeDisabled();
		expect(getByLabelText('Previous Page')).toBeDisabled();
		expect(getByLabelText('Next Page')).toBeDisabled();
		expect(getByLabelText('Last Page')).toBeDisabled();
	});

	it('should disable only first page and previous page buttons if currentPage is less than totalPages', () => {
		const { getByLabelText } = render(
			<Controls
				data={mockData}
				handlePagination={mockHandlePagination}
				currentPage={1}
				totalPages={2}
			/>,
		);

		expect(getByLabelText('First Page')).toBeDisabled();
		expect(getByLabelText('Previous Page')).toBeDisabled();

		expect(getByLabelText('Next Page')).toBeEnabled();
		expect(getByLabelText('Last Page')).toBeEnabled();
	});

	it('should disable only next page and last page buttons if currentPage equals totalPages', () => {
		const { getByLabelText } = render(
			<Controls
				data={mockData}
				handlePagination={mockHandlePagination}
				currentPage={2}
				totalPages={2}
			/>,
		);

		expect(getByLabelText('First Page')).toBeEnabled();
		expect(getByLabelText('Previous Page')).toBeEnabled();

		expect(getByLabelText('Next Page')).toBeDisabled();
		expect(getByLabelText('Last Page')).toBeDisabled();
	});

	it('should exec handlePagination passing 1 as argument when clicking first page button', () => {
		const { getByLabelText } = render(
			<Controls
				data={mockData}
				handlePagination={mockHandlePagination}
				currentPage={2}
				totalPages={2}
			/>,
		);

		fireEvent.click(getByLabelText('First Page'));

		expect(mockHandlePagination.mock.calls).toHaveLength(1);
		// The first argument of the first call to the function should be 1
		expect(mockHandlePagination.mock.calls[0][0]).toBe(1);
	});

	it('should exec handlePagination passing currentPage - 1 as argument when clicking previous page button', () => {
		const currentPage = 3;
		const { getByLabelText } = render(
			<Controls
				data={mockData}
				handlePagination={mockHandlePagination}
				currentPage={currentPage}
				totalPages={4}
			/>,
		);

		fireEvent.click(getByLabelText('Previous Page'));

		expect(mockHandlePagination.mock.calls).toHaveLength(1);
		// The first argument of the first call to the function should be 2
		expect(mockHandlePagination.mock.calls[0][0]).toBe(currentPage - 1);
	});

	it('should exec handlePagination passing currentPage + 1 as argument when clicking next page button ', () => {
		const currentPage = 1;
		const { getByLabelText } = render(
			<Controls
				data={mockData}
				handlePagination={mockHandlePagination}
				currentPage={currentPage}
				totalPages={3}
			/>,
		);

		fireEvent.click(getByLabelText('Next Page'));

		expect(mockHandlePagination.mock.calls).toHaveLength(1);
		// The first argument of the first call to the function should be 2
		expect(mockHandlePagination.mock.calls[0][0]).toBe(currentPage + 1);
	});

	it('should exec handlePagination passing 3 as argument when clicking last page button and totalPages equals 3', () => {
		const { getByLabelText } = render(
			<Controls
				data={mockData}
				handlePagination={mockHandlePagination}
				currentPage={1}
				totalPages={3}
			/>,
		);

		fireEvent.click(getByLabelText('Last Page'));

		expect(mockHandlePagination.mock.calls).toHaveLength(1);
		// The first argument of the first call to the function should be 3
		expect(mockHandlePagination.mock.calls[0][0]).toBe(3);
	});
});
