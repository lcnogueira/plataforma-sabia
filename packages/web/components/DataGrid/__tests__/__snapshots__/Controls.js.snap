// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Controls /> should render without crashing 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 button {
  color: hsla(0,0%,40%);
  background: none;
  border: none;
}

.c0 button:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0 .sort-by {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 .sort-by .sort-order {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .sort-by > div {
  margin-right: 1.2rem;
}

.c0 .sort-by > div > button:last-of-type {
  margin-left: 1.2rem;
}

.c0 .pagination {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: auto;
}

.c0 .pagination > span {
  font-size: 1.4rem;
}

.c0 .sort-by > div,
.c0 .pagination {
  margin-bottom: 1.4rem;
}

<div>
  <div
    class="c0"
  />
</div>
`;
